src/archivist/services/index.js

```diff
@@ -13,91 +14,104 @@ const declarationsPath = path.resolve(__dirname, '../../..', config.get('service
+  const jsonDeclarationFilePath = path.join(declarationsPath, `${serviceId}.json`);
+  const rawServiceDeclaration = await fs.readFile(jsonDeclarationFilePath);
+
+  try {
+    return JSON.parse(rawServiceDeclaration);
+  } catch (error) {
+    throw new Error(`The "${serviceId}" service declaration is malformed and cannot be parsed in ${jsonDeclarationFilePath}`);
+  }
+}
+
+async function loadServiceFilters(serviceId, filterNames) {
+  if (!filterNames) {
+    return;
+  }
+
+  const filterFilePath = `${serviceId}.filters.js`;
+  const serviceFilters = await import(pathToFileURL(path.join(declarationsPath, filterFilePath))); // eslint-disable-line no-await-in-loop
+
+  return filterNames.map(filterName => serviceFilters[filterName]);
+}
+
+async function loadServiceDocument(service, documentType, documentTypeDeclaration) {
+  const { filter: filterNames, fetch: location, executeClientScripts, select: contentSelectors, remove: noiseSelectors, combine } = documentTypeDeclaration;
+
+  const pages = [];
+
+  const filters = await loadServiceFilters(service.id, filterNames);
+
+  if (!combine) {
+    pages.push(new PageDeclaration({ location, executeClientScripts, contentSelectors, noiseSelectors, filters }));
+  } else {
+    for (const pageDeclaration of combine) {
+      const { filter: pageFilterNames, fetch: pageLocation, pageExecuteClientScripts, select: pageContentSelectors, remove: pageNoiseSelectors } = pageDeclaration;
+
+      const pageFilters = await loadServiceFilters(service.id, pageFilterNames); // eslint-disable-line no-await-in-loop
+
+      pages.push(new PageDeclaration({
+        location: pageLocation || location,
+        executeClientScripts: pageExecuteClientScripts || executeClientScripts,
+        contentSelectors: pageContentSelectors || contentSelectors,
+        noiseSelectors: pageNoiseSelectors || noiseSelectors,
+        filters: pageFilters || filters,
+      }));
+    }
+  }
+
+  service.addDocumentDeclaration(new DocumentDeclaration({ service, type: documentType, pages }));
+}
+
+async function getDeclaredServicesIds() {
+  const fileNames = await fs.readdir(declarationsPath);
+
+  const servicesFileNames = fileNames.filter(fileName => path.extname(fileName) == '.json' && !fileName.includes('.history.json'));
+
+  return servicesFileNames.map(serviceFileName => path.basename(serviceFileName, '.json'));
+}
+
 export async function loadWithHistory(servicesIds = []) {
   const services = await load(servicesIds);
 
   for (const serviceId of Object.keys(services)) {
-    const { declaration, filters } = await loadServiceHistoryFiles(serviceId); // eslint-disable-line no-await-in-loop
+    const { declarations, filters } = await loadServiceHistoryFiles(serviceId); // eslint-disable-line no-await-in-loop
 
-    for (const documentType of Object.keys(declaration)) {
-      const documenTypeDeclarationEntries = declaration[documentType];
+    for (const documentType of Object.keys(declarations)) {
```

Comment:
Why not `Object.entries(declarations).forEach(([ documentType, documentDeclaration ]) => {`? 🙂

Reply:
Because using a `forEach` instead of a native `for` loop won't ensure loading is effectively done before returning services.